### PR TITLE
Files patched for Issue# 2352 (unordered network replication of node hierarchies)

### PR DESCRIPTION
**Pull request by LeithKetchell**
Connection class was gathering dirty node IDs in an unordered container (HashMap), which causes network Clients to process newly-replicated nodes in an arbitrary order.
Resolved by replacing the HashMap with a Vector (an ordered container).

Please disregard the initial commit as my project files were not up to date.

### CODE REVIEW

**Comment by eugeneko** [comment]
The idea of using Vector isn't actually bad.
However, I'm unsire if it's working properly in current implementation.

https://github.com/urho3d/Urho3D/pull/2412/files#diff-7b9673cfc396212ff55f1c1bee79ab6eL1159

What if Scene::MarkReplicationDirty called twice for the same node? Is it possible?

https://github.com/urho3d/Urho3D/pull/2412/files#diff-55761705d5f88d7b0ec255a02b3d3ae1L1792

It looks similar, but in this line double-adding seems impossible. Is it true? Or node can appear in the dirty list without marking dirty?

https://github.com/urho3d/Urho3D/pull/2412/files#diff-b12ad27bb5ee6ae88c0b4c3766ead908L1136

Is it really ok to have quadratic node processing complexity?
Does this function really have to be called with `unsigned nodeID` parameter instead of index?

**Comment by LeithKetchell** [comment]
Hey, let's talk about it :)

I am assuming a few things, so let me start with my assumptions.

Assuming that we're single-threaded... by the time we process dirty nodes, they can't be re-marked as dirty - that part of the processing has been completed, and so they (should not?) appear twice in the list of dirty nodes... The changes I propose occur late in the piece, where we have our list of dirty nodes, presumably collected in a top-down fashion during evaluation of the scene node hierarchy (so parent nodes should appear prior to child nodes, irrespective of the possibly-recycled IDs).
I may be barking up the wrong tree, which is why this patch really needs to be tested, in the use-case that was proposed by the OP of #2352, which was the serialization of a node hierarchy whose IDs are not representative of the order of the node hierarchy... In other words, just create a few nodes, connect them in a different order than you created them, and see if they replicate correctly.
We know there is an issue for the master branch, let's see if my patch is enough to fix it.
If not, we're a step closer to a full patch, and can stare down the root cause.

**Comment by ArnisLielturks** [comment]
I want these changes to be added to the engine as I have experienced the problems caused without them. But has anyone actually tried to test them with the proposed scenario? https://github.com/urho3d/Urho3D/issues/2352

File: Source/Urho3D/Network/Connection.cpp [hunk 1]
```diff
@@ -146,7 +146,7 @@ void Connection::SendRemoteEvent(Node* node, StringHash eventType, bool inOrder,
         URHO3D_LOGERROR("Sender node is not in the connection's scene, can not send remote node event");
         return;
     }
-    if (!node->IsReplicated())
+    if (node->GetID() >= FIRST_LOCAL_ID)
```
Comment (PredatorMF) [inline Source/Urho3D/Network/Connection.cpp]: Can this line be commented or explained ?

File: Source/Urho3D/Network/Connection.cpp [hunk 2]
```diff
@@ -335,16 +325,16 @@ void Connection::SendRemoteEvents()
 
 void Connection::SendPackages()
 {
-    while (!uploads_.Empty())
+    while (!uploads_.Empty() && connection_->NumOutboundMessagesPending() < 1000)
```
Comment (PredatorMF) [inline Source/Urho3D/Network/Connection.cpp]: 1000 is a bit magical, what is the reason behind this number ?

File: Source/Urho3D/Network/Connection.cpp [hunk 3]
```diff
@@ -335,16 +325,16 @@ void Connection::SendRemoteEvents()
...
         unsigned char buffer[PACKAGE_FRAGMENT_SIZE];
 
         for (HashMap<StringHash, PackageUpload>::Iterator i = uploads_.Begin(); i != uploads_.End();)
         {
             HashMap<StringHash, PackageUpload>::Iterator current = i++;
             PackageUpload& upload = current->second_;
-            auto fragmentSize =
-                (unsigned)Min((int)(upload.file_->GetSize() - upload.file_->GetPosition()), (int)PACKAGE_FRAGMENT_SIZE);
+            unsigned fragmentSize =
+                Min((upload.file_->GetSize() - upload.file_->GetPosition()), PACKAGE_FRAGMENT_SIZE);
```
Comment (PredatorMF) [inline Source/Urho3D/Network/Connection.cpp]: The auto was fine.

File: Source/Urho3D/Network/Connection.cpp [hunk 4]
```diff
@@ -491,7 +470,7 @@ void Connection::ProcessLoadScene(int msgID, MemoryBuffer& msg)
 
     // In case we have joined other scenes in this session, remove first all downloaded package files from the resource system
     // to prevent resource conflicts
-    auto* cache = GetSubsystem<ResourceCache>();
+    ResourceCache* cache = GetSubsystem<ResourceCache>();
```
Comment (PredatorMF) [inline Source/Urho3D/Network/Connection.cpp]: Can't this be auto, C++11 is already a requirement and most of the code tries to follow.

File: Source/Urho3D/Network/Connection.cpp [hunk 5]
```diff
@@ -557,7 +536,7 @@ void Connection::ProcessSceneUpdate(int msgID, MemoryBuffer& msg)
 
             // Read initial attributes, then snap the motion smoothing immediately to the end
             node->ReadDeltaUpdate(msg);
-            auto* transform = node->GetComponent<SmoothedTransform>();
+            SmoothedTransform* transform = node->GetComponent<SmoothedTransform>();
```
Comment (PredatorMF) [inline Source/Urho3D/Network/Connection.cpp]: Please, auto.

File: Source/Urho3D/Network/Connection.cpp [hunk 6]
```diff
@@ -1245,7 +1218,7 @@ void Connection::ProcessExistingNode(Node* node, NodeReplicationState& nodeState
 
     // Check from the interest management component, if exists, whether should update
     /// \todo Searching for the component is a potential CPU hotspot. It should be cached
-    auto* priority = node->GetComponent<NetworkPriority>();
+    NetworkPriority* priority = node->GetComponent<NetworkPriority>();
```
Comment (PredatorMF) [inline Source/Urho3D/Network/Connection.cpp]: auto priority...

File: Source/Urho3D/Network/Connection.h [hunk 7]
```diff
@@ -201,7 +195,7 @@ class URHO3D_API Connection : public Object
     float GetRoundTripTime() const;
 
     /// Return the time since last received data from the remote host in milliseconds.
-    unsigned GetLastHeardTime() const;
+    float GetLastHeardTime() const;
```
Comment (PredatorMF) [inline Source/Urho3D/Network/Connection.h]: float is not cool for a timestamp or time deltas

**Comment by eugeneko** [comment]
@PredatorMF you are commenting outdated code

**Comment by PredatorMF** [comment]
Its possible, hard to navigate through my phone, I'll take those comments down.

**Comment by eugeneko** [comment]
@SirNate0 Linear search in list would be about 100 times slower than in vector, so I don't really think list is better in this case. List is good in only one case -- when you almost don't work with container as a whole (no loops) but only with individual elements.

Maybe we can do some specialized container combination...

**Pull request closed without merging**